apps/passport-client/src/validateState.ts

```diff
@@ -0,0 +1,158 @@
+import { User, requestLogToServer } from "@pcd/passport-interface";
+import { PCDCollection } from "@pcd/pcd-collection";
+import {
+  SemaphoreIdentityPCD,
+  SemaphoreIdentityPCDPackage
+} from "@pcd/semaphore-identity-pcd";
+import { Identity } from "@semaphore-protocol/identity";
+import { appConfig } from "./appConfig";
+import { loadSelf } from "./localstorage";
+
+/**
+ * Validates the application state using {@link validateAppState}. In the case
+ * that the state is invalid, returns `true`, and concurrently uploads the validation
+ * errors to the server for further inspection.
+ *
+ * In the case there are no validation errors, returns `false`.
+ */
+export function validateAndLogStateErrors(
+  self: User | undefined,
+  identity: Identity | undefined,
+  pcds: PCDCollection | undefined,
+  forceCheckPCDs?: boolean
+): boolean {
+  const validationErrors = validateAppState(
+    self,
+    identity,
+    pcds,
+    forceCheckPCDs
+  );
+
+  if (validationErrors.errors.length > 0) {
+    logValidationErrors(validationErrors);
+    return false;
+  }
+
+  return true;
+}
+
+/**
+ * Determines whether the app's global state contains valid data. If it does not,
+ * returns the set of things that are incorrect about it in a {@link ValidationErrors}
+ * object. If there were no validation errors the result will contain an empty array
+ * of errors.
+ *
+ * The provided {@link PCDCollection} is not checked unless either this function
+ * determines the user is logged in or the {@link forceCheckPCDs} argument is `true`.
+ */
+function validateAppState(
+  self: User | undefined,
+  identity: Identity | undefined,
+  pcds: PCDCollection | undefined,
+  forceCheckPCDs?: boolean
+): ValidationErrors {
+  const validationErrors: string[] = [];
+
+  const loggedOut = !self;
+
+  const identityPCDFromCollection = pcds?.getPCDsByType(
+    SemaphoreIdentityPCDPackage.name
+  )?.[0] as SemaphoreIdentityPCD | undefined;
```

Comment:
?.[0] returns undefined rather than crashing, I assume?  Not the clearest, but seems okay for this case.